### PR TITLE
Assign UUIDs to ships, missions, and npcs

### PR DESCRIPTION
**Pull request by NomadicVolcano**
**Feature:** Adds a unique identifier to ships, missions, and NPCs. Also adds a Random::UUID() function to generate random UUIDs. This aids in recreating pointers to missions, ships, and npcs after loading a game.

## Feature Details

UUIDs are these: https://en.wikipedia.org/wiki/Universally_unique_identifier
Standard definition: https://tools.ietf.org/html/rfc4122

In your save files, you should see:

```
ship
  ...
  uuid 2c8ab67c-e7cc-471a-a415-08a63ba52527

mission
  ...
  uuid 27ecb32b-e8a2-4350-aa5b-6b4b773520c7
  ...
  npc
    ...
    uuid b45551e8-324c-4b35-96ba-cb02357694e2
```

## Usage Examples

In and of itself, this PR is useless, but it will be needed by other PRs that have to uniquely identify ships, missions, or NPCs when loading data back from a file. This allows pointers to ships, NPCs, and missions to be restored.

Used by #5070 to load pointers to mission cargo and mission passengers, and #5082 to load the mission whose conversation was being displayed.

## Testing Done

1. Load a game with no UUIDs. Save it. Make sure all ships, missions, available missions, available jobs, and NPCs have UUIDs.

2. Buy a ship. Save. Make sure it has a UUID.

3. Go to a planet with available jobs and available missions. Save. Make sure they have UUIDs.

4. Go to a planet with a job or mission that has an NPC. Save. Make sure the NPC and its ships have UUIDs.

5. Run with this branch. Save. Load mission from earlier version of the code. It should load correctly; the uuids should be discarded.

## Performance Impact

It'll take a fraction of a microsecond longer to instantiate a mission or make a ship.

### CODE REVIEW

**Comment by NomadicVolcano** [comment]
**Someone with a windows or mac machine needs to test this.**

The code in the `Random` class that generates these has two lines that differ between Linux and non-Linux. I only have Linux, so I need someone else to test this on other platforms.

**Comment by Zitchas** [comment]
1. Will this compare the generated value with already existing values to ensure that the new uid is unique?
2. Will this include a way for missions to store the resulting uid in a way that can be referred to in subsequent missions?
ex. Mission 1 gives the player a ship and generates a uid.
Mission 2 requires the player to be flying that ship in order to trigger.
The `on complete` of mission 3 removes the ship from the player.

Basically, we need some way to either communicate the generated uid's between missions; or have some kind of a system to ensure that missions can assign specific uid's to ships so that content creators can refer to that specific ship in future missions. The former sounds like more work on the engine side; the later is more work on the creator side. As a creator, the former sounds awesome, but I'm fine with having to manually assign uid's for ships that I'm going to be referring to in future missions. If we go the later route, there should probably be some format for defining ranges of values that are restricted to manually assigned values. For instance, perhaps values where the first `2c8ab67c` block of the uid is equal to the first 8 letters of a government. `remnant0` for Remnant missions, `syndicat` for the syndicate, etc.

**Comment by NomadicVolcano** [comment]
> Will this compare the generated value with already existing values to ensure that the new uid is unique?

The idea behind UUID is that it is sufficiently large that you can't have collisions, so no such checks are needed. If we run into problems, there are smarter UUID-generation algorithms, but I'm using a 128-bit random number, which should be good enough. (Actually, 121 random bits and 7 bits of format definition.)

> Will this include a way for missions to store the resulting uid in a way that can be referred to in subsequent missions?

I suspect tags would be a better option for the use cases you describe.

The UUIDs are mainly to set up pointers correctly when saving and loading. For example, #5070 uses it to keep track of mission cargo in ship holds. 

We need string variables and a reflexive API for the UUIDs to be usable in-game. Once we have that though, there are a great many things that could be done.

File: source/Mission.cpp [hunk 1]
```diff
@@ -1107,6 +1130,7 @@ Mission Mission::Instantiate(const PlayerInfo &player, const shared_ptr<Ship> &b
 	
 	// Generate the substitutions map.
 	map<string, string> subs;
+	subs["<uuid>"] = result.uuid;
```
Comment (petervdmeer) [inline source/Mission.cpp:1133]: Do we ever want to display the UUID in-game?
I would expect the UUID not to show up on the UI anywhere.

File: source/Mission.h [hunk 2]
```diff
@@ -67,6 +68,9 @@ class Mission {
 	// offered if no other missions (minor or otherwise) are being offered.
 	bool IsMinor() const;
 	
+	// For backward-compatibility to old saves, generate a UUID if none is present
```
Comment (petervdmeer) [inline source/Mission.h:71]: The current ES style rules requires comment to end with a dot.

File: source/Random.cpp [hunk 3]
```diff
@@ -112,3 +116,36 @@ double Random::Normal()
 #endif
 	return normal(gen);
 }
+
+
+
+// Get a random UUID
```
Comment (petervdmeer) [inline source/Random.cpp:122]: Maybe add some text to the comment here how an UUID looks?
Also, comments in ES need to be full sentences and end with a dot.

File: source/Mission.cpp [hunk 4]
```diff
@@ -416,6 +430,14 @@ bool Mission::IsMinor() const
 
 
 
+void Mission::EnsureUUID()
+{
+	if(uuid.size() < 1)
```
Comment (petervdmeer) [inline source/Mission.cpp:435]: Most other sections (for ships and NPCs) use `if(uuid.empty())`. Is there a reason the check here is different?

File: source/Random.cpp (hunk 3 above)
Comment (NomadicVolcano) [reply source/Random.cpp]: I added an example, the IETF RFC number, and an explanation to both Random.cpp and Random.h.

File: source/Mission.cpp (hunk 1 above)
Comment (NomadicVolcano) [reply source/Mission.cpp]: Removed.

It was for debugging. I suppose if I need to debug again, I can put that line back in manually.

**Comment by NomadicVolcano** [comment]
I've fixed the problems @petervdmeer found.

**Comment by NomadicVolcano** [comment]
This PR is backward-compatible to old saves. The mission, npc, and ship Load() functions are smart enough to ignore the uuid.

File: source/Random.cpp [hunk 5]
```diff
@@ -27,10 +27,14 @@ namespace {
 	mt19937_64 gen;
 	uniform_int_distribution<uint32_t> uniform;
 	uniform_real_distribution<double> real;
+	uniform_real_distribution<> zeroToFifteen(0, 15);
+	uniform_real_distribution<> eightToEleven(8, 11);
 #else
 	thread_local mt19937_64 gen;
 	thread_local uniform_int_distribution<uint32_t> uniform;
 	thread_local uniform_real_distribution<double> real;
+	thread_local uniform_real_distribution<> zeroToFifteen(0, 15);
```
Comment (tehhowch) [inline source/Random.cpp:36]: Specify the type

File: source/Random.cpp [hunk 6]
```diff
@@ -112,3 +116,38 @@ double Random::Normal()
...
+// Get a version 4 (random) Universally Unique Identifier (see IETF RFC 4122)
+// which are hexidecimal strings like "2c8ab67c-e7cc-471a-a415-08a63ba52527"
+// and contain 121 bits of random data.
+string Random::UUID()
+{
+#ifndef __linux__
+	lock_guard<mutex> lock(workaroundMutex);
+#endif
+	char uuid[37];
+	static const char * const hex="0123456789abcdef";
```
Comment (tehhowch) [inline source/Random.cpp:131]: Space your operators.

File: source/Random.cpp [hunk 7]
```diff
@@ -112,3 +116,38 @@ double Random::Normal()
...
+	int i=0;
+	
+	for(int j = 0; j < 8; j++)
```
Comment (tehhowch) [inline source/Random.cpp:134]: prefix over postfix.

**Comment by NomadicVolcano** [comment]
> UUID class that manages the generating the UUID on first use

Yes, that thought occurred to me as well in recent days. I'll implement it when I catch up with my other PRs. I want to finish the Korath Exiles expansion first.

**Comment by NomadicVolcano** [comment]
Closing all pull requests related to this account.

**Comment by tatami4** [comment]
> create a UUID class that manages the generating the UUID on first use.

@tehhowch  What do you mean by this? Could you explain it more? Right now I don't see how this "UUID" class would work. Who would call it to generate the UUIDs?

**Comment by Zitchas** [comment]
I can't speak for the coding side, but conceptually, when a mission gives the player a ship, it would do one of two things:
a) assign the ship a specific preset UID (this would be good for unique ships), or
b) assign the ship a freshly generated UID and record it somewhere. (this would be good for times when a future mission wants to know that the player is using the ship it gave earlier, but that ship isn't necessarily unique in and of itself. For instance, if someone hired the player to pilot ship XYZ ship from point A to B, the mission would want to be able to check that the player is still flying ship XYZ. This could be a generic job with specific ships for that instance, but not important in the larger scheme of things)
c) When the player purchases a ship, it would be assigned a randomly generated UID. useful for when a mission wants to make sure the player is using a specific ship. Say, if they won some famous battle flying the Crimson Flee, and someone wants to hire them to fly the ship they won the battle with in a parade somewhere. This could be used for that.

**Comment by tehhowch** [comment]
I've implemented UUIDs in the sequence of commits:
954f6eb7ff414349b0417822e530352988060b60 (this PR's foundational work)
eff16d4c03c99dc45663d29bba9d866fd457e7da (my request for lazy initialization)
497722c9a23709b7b87884c3802a9fb58d37bb4b (use of platform-specific APIs instead of an oversized `std::string`)

**Comment by Zitchas** [comment]
Awesome!

**Pull request closed without merging**